bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java

```diff
@@ -0,0 +1,152 @@
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.knx.internal.discovery;
+
+import static org.openhab.binding.knx.internal.KNXBindingConstants.THING_TYPE_IP_BRIDGE;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.Future;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.config.discovery.AbstractDiscoveryService;
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.config.discovery.DiscoveryService;
+import org.openhab.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import tuwien.auto.calimero.knxnetip.Discoverer;
+import tuwien.auto.calimero.knxnetip.Discoverer.Result;
+import tuwien.auto.calimero.knxnetip.servicetype.SearchResponse;
+import tuwien.auto.calimero.knxnetip.util.ServiceFamiliesDIB;
+import tuwien.auto.calimero.knxnetip.util.ServiceFamiliesDIB.ServiceFamily;
+
+/**
+ * Discovers KNXnet/IP interfaces or routers and adds the results to the inbox.
+ * Several items per device might be created, as routers typically support routing and tunneling.
+ * Discovery uses multicast traffic to IP 224.0.23.12, port 3671.
+ *
+ * @implNote Discovery is based on the functionality provided by Calimero library.
+ * @author Holger Friedrich - Initial contribution
+ */
+@Component(service = DiscoveryService.class, configurationPid = "discovery.knx")
+@NonNullByDefault
+public class KNXnetDiscoveryService extends AbstractDiscoveryService {
+    private final Logger logger = LoggerFactory.getLogger(KNXnetDiscoveryService.class);
+
+    private @Nullable Future<?> scanFuture = null;
+
+    public void activate() {
+        super.activate(null);
+        startScan();
+    }
+
+    @Override
+    public void deactivate() {
+        stopScan();
+        super.deactivate();
+    }
+
+    @Activate
+    public KNXnetDiscoveryService() {
+        super(Set.of(THING_TYPE_IP_BRIDGE), 3, true);
+    }
+
+    @Override
+    protected void startScan() {
+        if (scanFuture == null) {
+            scanFuture = scheduler.submit(this::startDiscovery);
+        } else {
+            logger.debug("KNXnet/IP background discovery scan in progress");
+        }
+    }
+
+    @Override
+    protected void stopScan() {
+        if (scanFuture != null) {
+            scanFuture.cancel(false);
+            scanFuture = null;
+        }
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        // only start once at startup
+        startScan();
```

Comment:
You already start it in `activate()`.
The "background discovery" is meant to do a continuous scan in the background, so that the Inbox is filled even if the user does not trigger a scan. So I would rather expect a scheduled job here.

Reply:
removed

Reply:
Removing it means you are now not reacting to a configuration change of "background discovery enabled" anymore - i.e. you are not supporting any background discovery.
You have two options now:
1. Either call `super()` in the constructor with `false` instead of `true` now or
2. schedule a recurring job here (and unschedule it again in `stopBackgroundDiscovery()`).

Reply:
What I want to archive is basically:
- at startup scan once (unless disabled)
- maybe start scanning if user switches from disabled to enabled
- at best have it triggerable from the OH console: ```discovery start knx```

I think a found the way using the methods from ```AbstractDiscoveryService```. Seems to work as intended.

Reply:
This means you do not want to support background discovery, i.e. you choose option (1) from above.
As a result, you should call `super` with `false` [here](https://github.com/openhab/openhab-addons/pull/16033/files#diff-68ecbf945b85d60351bc4142da436b4dc4d3a980e045763ac756cbcaeeba7c00R55) and not implement anything in the start/stop background discovery methods.

Reply:
Hm, I think the current implementation is a bit of both: I have done it as background discovery, but the scan is only running once.
The pro of this approach is:
- it can be completely disabled using ```discovery.knx:background=false```
- it can be triggered by ```discovery start knx```
- and it is running only once at startup

The alternative as you outlined above, use ```super(..., false)``` and start the scan from the ctor. Then we users cannot disable it.

Reply:
Hm, ok, I see your point. An IP interface is unlikely to appear suddenly, so having a single scan at startup is usually enough and it is convenient for the user to find it in the inbox without having to manually trigger a scan for KNX things. So alright, let's go for this.